<h1 align='center'> CSC4120 Programming Assignment 1 </h1>

## Submission Requirements

   The submission <font color = #942F2C>deadline is Sep 21 (Sun.), 2025, 11:59 pm</font>. Solutions submitted after the deadline will be graded as 0 points. Please submit an **ipynb** file and clearly state your group members' student IDs. Otherwise, your points will be deducted.

## What you need to do

1. Understand the document distance problem.

2. Understand the python code and how we improve the algorithm in each step.

3. Implement merge sort and the dictionary version.

## Student IDs
Group13

122090437 QIU Runheng
122090669 YE Shuhuan
122090872 JIN Yiyao

\###

In [11]:
import math
import sys
import cProfile
import string

filename_1 = "file1.txt"
filename_2 = "file2.txt"
translation_table = str.maketrans(string.punctuation + string.ascii_uppercase,
                                     " "*len(string.punctuation) + string.ascii_lowercase)

for k, v in translation_table.items():
    if not chr(k).isalnum():
        print(f"{chr(k)} ({k}) -> {v}")


! (33) -> 32
" (34) -> 32
# (35) -> 32
$ (36) -> 32
% (37) -> 32
& (38) -> 32
' (39) -> 32
( (40) -> 32
) (41) -> 32
* (42) -> 32
+ (43) -> 32
, (44) -> 32
- (45) -> 32
. (46) -> 32
/ (47) -> 32
: (58) -> 32
; (59) -> 32
< (60) -> 32
= (61) -> 32
> (62) -> 32
? (63) -> 32
@ (64) -> 32
[ (91) -> 32
\ (92) -> 32
] (93) -> 32
^ (94) -> 32
_ (95) -> 32
` (96) -> 32
{ (123) -> 32
| (124) -> 32
} (125) -> 32
~ (126) -> 32


## 1. Initial version of document distance

This program computes the "distance" between two text files as the angle between their word frequency vectors (in radians).

For each input file, a word-frequency vector is computed as follows:

   (1) the specified file is read in

   (2) it is converted into a list of alphanumeric "words"

       Here a "word" is a sequence of consecutive alphanumeric
       characters.  Non-alphanumeric characters are treated as blanks.
       Case is not significant.

   (3) for each word, its frequency of occurrence is determined

The "distance" between two vectors is the angle between them.

If $ x = (x_1, x_2, ..., x_n) $ is the first vector ($ x_i $ = freq of word i)
and $ y = (y_1, y_2, ..., y_n) $ is the second vector,
then the angle between them is defined as:

   $ d(x,y) = \arccos{\frac{innerProduct(x,y)}{norm(x)*norm(y)}} $

where:

   $ innerProduct(x,y) = x_1*y_1 + x_2*y_2 + \cdots + x_n*y_n $

   $ norm(x) = \sqrt{innerProduct(x,x)} $

   ***


### What you need to do

Run the code and report the running time.

\###

Your answer goes here

\###
The distance between the documents is: 0.619328 radians
| Function | Calls | Total Time (s) | Cumulative Time (s) | Per Call (s) |
|----------|-------|----------------|-------------------|--------------|
| interactiveshell.py:3543(run_code) | 2 | 0.000 | 1.866 | 0.000 |
| {built-in method builtins.exec} | 2 | 0.000 | 1.866 | 0.000 |
| 3696036711.py:1(<module>) | 1 | 0.000 | 1.866 | 0.000 |
| 2036849071.py:85(docdist1) | 1 | 0.001 | 1.866 | 0.001 |
| 2036849071.py:51(word_frequencies_for_file) | 2 | 0.000 | 1.847 | 0.000 |
| 2036849071.py:13(get_words_from_line_list) | 2 | 1.552 | 1.567 | 0.776 |
| 2036849071.py:37(count_frequency) | 2 | 0.277 | 0.277 | 0.139 |
| 2036849071.py:75(vector_angle) | 1 | 0.000 | 0.018 | 0.000 |
| 2036849071.py:60(inner_product) | 3 | 0.018 | 0.018 | 0.006 |
| 2036849071.py:24(get_words_from_string) | 23110 | 0.003 | 0.015 | 0.000 |
| {method 'translate' of 'str' objects} | 23110 | 0.009 | 0.009 | 0.000 |
| {method 'split' of 'str' objects} | 23110 | 0.003 | 0.003 | 0.000 |
| 2036849071.py:1(read_file) | 2 | 0.000 | 0.002 | 0.000 |
| {method 'readlines' of '_io._IOBase' objects} | 2 | 0.002 | 0.002 | 0.001 |
| codecs.py:319(decode) | 84 | 0.000 | 0.000 | 0.000 |
| interactiveshell.py:315(_modified_open) | 2 | 0.000 | 0.000 | 0.000 |
| {built-in method io.open} | 2 | 0.000 | 0.000 | 0.000 |
| {built-in method _codecs.utf_8_decode} | 84 | 0.000 | 0.000 | 0.000 |
| {built-in method builtins.print} | 1 | 0.000 | 0.000 | 0.000 |
| iostream.py:655(write) | 2 | 0.000 | 0.000 | 0.000 |
| iostream.py:577(_schedule_flush) | 2 | 0.000 | 0.000 | 0.000 |
| iostream.py:259(schedule) | 1 | 0.000 | 0.000 | 0.000 |
| {method 'append' of 'list' objects} | 1047 | 0.000 | 0.000 | 0.000 |
| socket.py:623(send) | 1 | 0.000 | 0.000 | 0.000 |
| ipkernel.py:781(_clean_thread_parent_frames) | 2 | 0.000 | 0.000 | 0.000 |
| codeop.py:117(__call__) | 2 | 0.000 | 0.000 | 0.000 |
| {built-in method builtins.compile} | 2 | 0.000 | 0.000 | 0.000 |
| ipkernel.py:796(<setcomp>) | 1 | 0.000 | 0.000 | 0.000 |
| threading.py:1145(ident) | 8 | 0.000 | 0.000 | 0.000 |
| traitlets.py:676(__get__) | 2 | 0.000 | 0.000 | 0.000 |
| threading.py:1478(enumerate) | 1 | 0.000 | 0.000 | 0.000 |
| contextlib.py:279(helper) | 2 | 0.000 | 0.000 | 0.000 |
| traitlets.py:629(get) | 2 | 0.000 | 0.000 | 0.000 |
| threading.py:1169(is_alive) | 1 | 0.000 | 0.000 | 0.000 |
| codecs.py:309(__init__) | 2 | 0.000 | 0.000 | 0.000 |
| contextlib.py:102(__init__) | 2 | 0.000 | 0.000 | 0.000 |
| contextlib.py:130(__enter__) | 2 | 0.000 | 0.000 | 0.000 |
| iostream.py:505(parent_header) | 2 | 0.000 | 0.000 | 0.000 |
| {built-in method builtins.next} | 4 | 0.000 | 0.000 | 0.000 |
| iostream.py:550(_is_master_process) | 2 | 0.000 | 0.000 | 0.000 |
| contextlib.py:139(__exit__) | 2 | 0.000 | 0.000 | 0.000 |
| compilerop.py:180(extra_flags) | 4 | 0.000 | 0.000 | 0.000 |
| threading.py:1102(_wait_for_tstate_lock) | 1 | 0.000 | 0.000 | 0.000 |
| iostream.py:138(_event_pipe) | 1 | 0.000 | 0.000 | 0.000 |
| interactiveshell.py:3495(compare) | 2 | 0.000 | 0.000 | 0.000 |
| codecs.py:260(__init__) | 2 | 0.000 | 0.000 | 0.000 |
| {built-in method builtins.getattr} | 4 | 0.000 | 0.000 | 0.000 |
| {method 'acquire' of '_thread.lock' objects} | 1 | 0.000 | 0.000 | 0.000 |
| {method '__exit__' of '_thread.RLock' objects} | 3 | 0.000 | 0.000 | 0.000 |
| {built-in method math.sqrt} | 1 | 0.000 | 0.000 | 0.000 |
| {built-in method posix.getpid} | 2 | 0.000 | 0.000 | 0.000 |
| {method 'get' of '_contextvars.ContextVar' objects} | 2 | 0.000 | 0.000 | 0.000 |
| {built-in method math.acos} | 1 | 0.000 | 0.000 | 0.000 |
| interactiveshell.py:1277(user_global_ns) | 2 | 0.000 | 0.000 | 0.000 |
| {built-in method builtins.isinstance} | 4 | 0.000 | 0.000 | 0.000 |
| {built-in method builtins.len} | 2 | 0.000 | 0.000 | 0.000 |
| {method 'items' of 'dict' objects} | 2 | 0.000 | 0.000 | 0.000 |
| {method 'values' of 'dict' objects} | 2 | 0.000 | 0.000 | 0.000 |
| {method 'write' of '_io.StringIO' objects} | 2 | 0.000 | 0.000 | 0.000 |
| typing.py:1737(cast) | 4 | 0.000 | 0.000 | 0.000 |
| threading.py:553(is_set) | 1 | 0.000 | 0.000 | 0.000 |
| {method 'keys' of 'dict' objects} | 4 | 0.000 | 0.000 | 0.000 |
| {method 'disable' of '_lsprof.Profiler' objects} | 1 | 0.000 | 0.000 | 0.000 |
| {method 'append' of 'collections.deque' objects} | 1 | 0.000 | 0.000 | 0.000 |

In [12]:
def read_file(filename):
    """ 
    Read the text file with the given filename;
    return a list of the lines of text in the file.
    """
    try:
        f = open(filename, 'r', encoding= 'utf-8')
        return f.readlines()
    except IOError:
        print("Error opening or reading input file: ", filename)
        sys.exit()

def get_words_from_line_list(L):
    """
    Parse the given list L of text lines into words.
    Return list of all words found.
    """
    word_list = []
    for line in L:
        words_in_line = get_words_from_string(line)
        word_list = word_list + words_in_line
    return word_list

def get_words_from_string(line):
    """
    Return a list of the words in the given input string,
    converting each word to lower-case.

    Input:  line (a string)
    Output: a list of strings 
              (each string is a sequence of alphanumeric characters)
    """
    line = line.translate(translation_table)
    word_list = line.split()
    return word_list

def count_frequency(word_list):
    """
    Return a list giving pairs of form: (word,frequency)
    """
    L = []
    for new_word in word_list:
        for entry in L:
            if new_word == entry[0]:
                entry[1] = entry[1] + 1
                break
        else:
            L.append([new_word, 1])
    return L

def word_frequencies_for_file(filename):
    """
    Return list of (word,frequency) pairs for the given file.
    """
    line_list = read_file(filename)
    word_list = get_words_from_line_list(line_list)
    freq_mapping = count_frequency(word_list)
    return freq_mapping

def inner_product(L1, L2):
    """
    Inner product between two vectors, where vectors
    are represented as lists of (word,freq) pairs.

    Example: inner_product([["and",3],["of",2],["the",5]],
                           [["and",4],["in",1],["of",1],["this",2]]) = 14.0 
    """
    sum = 0.0
    for word1, count1 in L1:
        for word2, count2 in L2:
            if word1 == word2:
                sum += count1 * count2
    return sum

def vector_angle(L1, L2):
    """
    The input is a list of (word,freq) pairs.

    Return the angle between these two vectors.
    """
    numerator = inner_product(L1, L2)
    denominator = math.sqrt(inner_product(L1, L1) * inner_product(L2, L2))
    return math.acos(numerator / denominator)

def docdist1():
    document_vector_1 = word_frequencies_for_file(filename_1)
    document_vector_2 = word_frequencies_for_file(filename_2)
    distance = vector_angle(document_vector_1, document_vector_2)
    print(f"The distance between the documents is: {distance:0.6f} radians")

cProfile.run("docdist1()")

The distance between the documents is: 0.619328 radians
         70646 function calls in 1.887 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    0.006    0.003 2036849071.py:1(read_file)
        2    1.570    0.785    1.588    0.794 2036849071.py:13(get_words_from_line_list)
    23110    0.004    0.000    0.017    0.000 2036849071.py:24(get_words_from_string)
        2    0.275    0.137    0.275    0.137 2036849071.py:37(count_frequency)
        2    0.000    0.000    1.869    0.934 2036849071.py:51(word_frequencies_for_file)
        3    0.018    0.006    0.018    0.006 2036849071.py:60(inner_product)
        1    0.000    0.000    0.018    0.018 2036849071.py:75(vector_angle)
        1    0.001    0.001    1.887    1.887 2036849071.py:85(docdist1)
        1    0.000    0.000    1.887    1.887 <string>:1(<module>)
        2    0.000    0.000    0.000    0.000 codecs.py:260(__init__)
        2  

In [13]:
import cProfile
import pstats
import io

# 1. 使用 cProfile 收集数据到一个字符串缓冲区
pr = cProfile.Profile()
pr.enable()
docdist1()
pr.disable()

s = io.StringIO()
ps = pstats.Stats(pr, stream=s)
ps.strip_dirs()
ps.sort_stats('cumtime')  # 按累计时间排序
ps.print_stats()  # 输出到 s 中

# 2. 解析输出，生成 Markdown 表格
lines = s.getvalue().splitlines()
md_lines = ["| Function | Calls | Total Time (s) | Cumulative Time (s) | Per Call (s) |",
            "|----------|-------|----------------|-------------------|--------------|"]

for line in lines[5:]:  # 前几行是表头，略过
    parts = line.split()
    if len(parts) >= 5:
        # pstats 输出格式: ncalls  tottime  percall  cumtime  percall  filename:lineno(function)
        calls = parts[0]
        tottime = parts[1]
        cumtime = parts[3]
        func_name = " ".join(parts[5:])
        per_call = parts[2]  # 每次调用的平均时间
        md_lines.append(f"| {func_name} | {calls} | {tottime} | {cumtime} | {per_call} |")

# 3. 输出 Markdown 表格
markdown_table = "\n".join(md_lines)
print(markdown_table)

The distance between the documents is: 0.619328 radians
| Function | Calls | Total Time (s) | Cumulative Time (s) | Per Call (s) |
|----------|-------|----------------|-------------------|--------------|
| interactiveshell.py:3543(run_code) | 2 | 0.000 | 1.866 | 0.000 |
| {built-in method builtins.exec} | 2 | 0.000 | 1.866 | 0.000 |
| 3696036711.py:1(<module>) | 1 | 0.000 | 1.866 | 0.000 |
| 2036849071.py:85(docdist1) | 1 | 0.001 | 1.866 | 0.001 |
| 2036849071.py:51(word_frequencies_for_file) | 2 | 0.000 | 1.847 | 0.000 |
| 2036849071.py:13(get_words_from_line_list) | 2 | 1.552 | 1.567 | 0.776 |
| 2036849071.py:37(count_frequency) | 2 | 0.277 | 0.277 | 0.139 |
| 2036849071.py:75(vector_angle) | 1 | 0.000 | 0.018 | 0.000 |
| 2036849071.py:60(inner_product) | 3 | 0.018 | 0.018 | 0.006 |
| 2036849071.py:24(get_words_from_string) | 23110 | 0.003 | 0.015 | 0.000 |
| {method 'translate' of 'str' objects} | 23110 | 0.009 | 0.009 | 0.000 |
| {method 'split' of 'str' objects} | 23110 | 0.003 | 

## 2. Change concatenate to extend in get_words_from_line_list

#### <span style="color:#f7651b">Compare the running time, analyze why we get improvement (or why not), identify it using cProfile.</span>

``` python
def get_words_from_line_list(L):
    """
    Parse the given list L of text lines into words.
    Return list of all words found.
    """
    word_list = []
    for line in L:
        words_in_line = get_words_from_string(line)
        word_list = word_list + words_in_line
    return word_list
```

In the previous implementation, `(get_words_from_line_list)` has a total running time of 1.570s and cumulative running time of 1.588s, which is the main bottleneck.

``` python
def get_words_from_line_list(L):
    """
    Parse the given list L of text lines into words.
    Return list of all words found.
    """
    word_list = []
    for line in L:
        words_in_line = get_words_from_string(line)
        word_list.extend(words_in_line)
    return word_list
```
In the new implementation, with the same similarity radians of $0.619328$, we have a total 93735 function calls in 0.328 seconds.
`(get_words_from_line_list)` has a total running time of 0.004s and cumulative running time of 0.020s, which is a great optimization.

The main reason for this would be the cost for `copy and paste`.

In the original code, we use naive `word_list = word_list + words_in_line`, during which os creates a new list instance, copy the original `word_list` and `words_in_line` into this new instance. In accumulation, this would cost $\mathcal{O}(N^2 \cdot M)$ if there are $N$ lines and $M$ words per line for example.

We look into the `cpython` library implementation( https://github.com/python/cpython/blob/main/Objects/listobject.c) of `list.extend()`: `static int
list_extend_fast(PyListObject *self, PyObject *iterable)`. And we can see that it first changes the iterable object into array, and add the new element in a <span style="color:#FF0000">dynamic array</span> mode. If there are free space (by checking $\frac{\text{list\_preallocate\_exact}}{\text{list\_resize}}$ ), we extend by inserting new element in an time complexity of $\mathcal{O}(1)$, and total time would be $\mathcal{O}(N_{\text{new\_elements}})$. If there are not enough space, we first dynamically enlarge the size, then do the same $\mathcal{O}(1)$ insertion.

``` c
static int
list_extend_fast(PyListObject *self, PyObject *iterable)
{
    Py_ssize_t n = PySequence_Fast_GET_SIZE(iterable);
    if (n == 0) {
        /* short circuit when iterable is empty */
        return 0;
    }

    Py_ssize_t m = Py_SIZE(self);
    // It should not be possible to allocate a list large enough to cause
    // an overflow on any relevant platform.
    assert(m < PY_SSIZE_T_MAX - n);
    if (self->ob_item == NULL) {
        if (list_preallocate_exact(self, n) < 0) {
            return -1;
        }
        Py_SET_SIZE(self, n);
    }
    else if (list_resize(self, m + n) < 0) {
        return -1;
    }

    // note that we may still have self == iterable here for the
    // situation a.extend(a), but the following code works
    // in that case too.  Just make sure to resize self
    // before calling PySequence_Fast_ITEMS.
    //
    // populate the end of self with iterable's items.
    PyObject **src = PySequence_Fast_ITEMS(iterable);
    PyObject **dest = self->ob_item + m;
    for (Py_ssize_t i = 0; i < n; i++) {
        PyObject *o = src[i];
        FT_ATOMIC_STORE_PTR_RELEASE(dest[i], Py_NewRef(o));
    }
    return 0;
}

```

Therefore, in the new code, `extend` add elements from `words_in_line` in place without creating new list instance and copy old elements. The time complexity would be $\mathcal{O}(N \cdot M)$ compare to original $\mathcal{O}(N^2 \cdot M)$, where $N$ denotes the number of lines, $M$ denotes the average words per line.


#### <span style="color:#f7651b">Your answer goes here:</span>
The distance between the documents is: 0.619328 (radians)




In [14]:
def get_words_from_line_list(L):
    """
    Parse the given list L of text lines into words.
    Return list of all words found.
    """
    word_list = []
    for line in L:
        words_in_line = get_words_from_string(line)
        word_list.extend(words_in_line)
    return word_list

def docdist2():
    document_vector_1 = word_frequencies_for_file(filename_1)
    document_vector_2 = word_frequencies_for_file(filename_2)
    distance = vector_angle(document_vector_1, document_vector_2)
    print("The distance between the documents is: %0.6f (radians)"%distance)

cProfile.run("docdist2()")

The distance between the documents is: 0.619328 (radians)
         93735 function calls in 0.327 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    0.005    0.003 2036849071.py:1(read_file)
    23110    0.004    0.000    0.019    0.000 2036849071.py:24(get_words_from_string)
        2    0.279    0.140    0.279    0.140 2036849071.py:37(count_frequency)
        2    0.000    0.000    0.308    0.154 2036849071.py:51(word_frequencies_for_file)
        3    0.018    0.006    0.018    0.006 2036849071.py:60(inner_product)
        1    0.000    0.000    0.018    0.018 2036849071.py:75(vector_angle)
        2    0.004    0.002    0.024    0.012 698513419.py:1(get_words_from_line_list)
        1    0.001    0.001    0.327    0.327 698513419.py:12(docdist2)
        1    0.000    0.000    0.327    0.327 <string>:1(<module>)
        2    0.000    0.000    0.000    0.000 codecs.py:260(__init__)
        2   

## 3. Sort the document vector
#### <span style="color:#f7651b">Compare the running time, analyze why we get improvement (or why not), identify it using cProfile.</span>

Time comparison between the original
The distance between the documents is: 0.619328 radians
| Implementation | Calls | Total Time (s) |
|----------|-------|----------------|
| naive concatenation | 70646 | 1.887 |
| list.extend method | 93735 | 0.327 |
| sort the document vector | 97501 | 0.318 |

#### <span style="color:#f7651b">Your answer goes here:</span>
Suppose $\text{document1}$ has $N1$ words with repetition, $\text{document2}$ has $N2$ words with repetition.

In either the naive concatenation or `list.extend()` implementation, we iterate through the `word_list` each time we calculate the word frequency of a document. Then we use this frequency to calculate the inner product. There are duplicate scanning. The time complexity for this would be $\mathcal{O}(N1 \times N2)$

In the new implementation, the general idea is to save time from duplicate scanning of previous words from word_list. So we first sort the words from incoming documents such that we can go in a straight-line-manner (no turning back) when calculating the inner product. The time complexity for this would be $\mathcal{O}(N1 \times N2)$


In [17]:
def insertion_sort(A):
    """
    Sort list A into order, in place.

    From Cormen/Leiserson/Rivest/Stein,
    Introduction to Algorithms (second edition), page 17,
    modified to adjust for fact that Python arrays use 
    0-indexing.
    """
    for j in range(len(A)):
        key = A[j]
        # insert A[j] into sorted sequence A[0..j-1]
        i = j - 1
        while i > -1 and A[i] > key:
            A[i + 1] = A[i]
            i = i - 1
        A[i + 1] = key
    return A
    
def word_frequencies_for_file(filename):
    """
    Return alphabetically sorted list of (word,frequency) pairs 
    for the given file.
    """
    line_list = read_file(filename)
    word_list = get_words_from_line_list(line_list)
    freq_mapping = count_frequency(word_list)
    insertion_sort(freq_mapping)
    return freq_mapping

def inner_product(L1, L2):
    """
    Inner product between two vectors, where vectors
    are represented as alphabetically sorted (word,freq) pairs.

    Example: inner_product([["and",3],["of",2],["the",5]],
                           [["and",4],["in",1],["of",1],["this",2]]) = 14.0 
    """
    sum = 0.0
    i = 0
    j = 0
    while i < len(L1) and j < len(L2):
        # L1[i:] and L2[j:] yet to be processed
        if L1[i][0] == L2[j][0]:
            # both vectors have this word
            sum += L1[i][1] * L2[j][1]
            i += 1
            j += 1
        elif L1[i][0] < L2[j][0]:
            # word L1[i][0] is in L1 but not L2
            i += 1
        else:
            # word L2[j][0] is in L2 but not L1
            j += 1
    return sum

def docdist3():
    sorted_word_list_1 = word_frequencies_for_file(filename_1)
    sorted_word_list_2 = word_frequencies_for_file(filename_2)
    distance = vector_angle(sorted_word_list_1, sorted_word_list_2)
    print("The distance between the documents is: %0.6f (radians)"%distance)

cProfile.run("docdist3()")

The distance between the documents is: 0.619328 (radians)
         97501 function calls in 0.318 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    0.005    0.002 2036849071.py:1(read_file)
    23110    0.004    0.000    0.017    0.000 2036849071.py:24(get_words_from_string)
        2    0.276    0.138    0.276    0.138 2036849071.py:37(count_frequency)
        1    0.000    0.000    0.000    0.000 2036849071.py:75(vector_angle)
        2    0.014    0.007    0.014    0.007 3836028585.py:1(insertion_sort)
        2    0.000    0.000    0.317    0.159 3836028585.py:20(word_frequencies_for_file)
        3    0.000    0.000    0.000    0.000 3836028585.py:31(inner_product)
        1    0.001    0.001    0.318    0.318 3836028585.py:57(docdist3)
        2    0.004    0.002    0.022    0.011 698513419.py:1(get_words_from_line_list)
        1    0.000    0.000    0.318    0.318 <string>:1(<module>)
   

## 4. Change sorting from insertion sort to merge sort

Implement merge sort.

Compare the running time, analyze why we get improvement (or why not), identify it using cProfile.

\###

Your answer goes here

\###

In [18]:
def merge(left, right):
    """
    Merge two sorted lists into a single sorted list.
    """
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i][0] <= right[j][0]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result

def merge_sort(A):
    """
    Sort list A into order, and return result.
    """
    #######################
    #                     #
    #     TO IMPLEMENT    #
    #                     #
    #######################
    if len(A) <= 1:
        return A

    mid = len(A) // 2
    left = merge_sort(A[:mid])
    right = merge_sort(A[mid:])

    return merge(left, right)

def word_frequencies_for_file(filename):
    """
    Return alphabetically sorted list of (word,frequency) pairs 
    for the given file.
    """
    line_list = read_file(filename)
    word_list = get_words_from_line_list(line_list)
    freq_mapping = count_frequency(word_list)
    freq_mapping = merge_sort(freq_mapping)
    return freq_mapping

def docdist3():
    sorted_word_list_1 = word_frequencies_for_file(filename_1)
    sorted_word_list_2 = word_frequencies_for_file(filename_2)
    distance = vector_angle(sorted_word_list_1, sorted_word_list_2)
    print("The distance between the documents is: %0.6f (radians)"%distance)

cProfile.run("docdist3()")

The distance between the documents is: 0.619328 (radians)
         132004 function calls (129914 primitive calls) in 0.310 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1045    0.002    0.000    0.002    0.000 1338857923.py:1(merge)
   2092/2    0.000    0.000    0.003    0.001 1338857923.py:21(merge_sort)
        2    0.000    0.000    0.308    0.154 1338857923.py:39(word_frequencies_for_file)
        1    0.001    0.001    0.310    0.310 1338857923.py:50(docdist3)
        2    0.000    0.000    0.005    0.003 2036849071.py:1(read_file)
    23110    0.004    0.000    0.018    0.000 2036849071.py:24(get_words_from_string)
        2    0.277    0.138    0.277    0.138 2036849071.py:37(count_frequency)
        1    0.000    0.000    0.000    0.000 2036849071.py:75(vector_angle)
        3    0.000    0.000    0.000    0.000 3836028585.py:31(inner_product)
        2    0.004    0.002    0.023    0.012 698513419.py:1(get_

## 5. Use dictionaries instead of lists

Implement the algorithm using dictionaries instead of lists. 

Analyze why we get improvement and identify it using cProfile.

\### 

Your answer goes here

\###

In [19]:
def read_file(filename):
    """ 
    Read the text file with the given filename;
    return a list of the lines of text in the file.
    """
    try:
        f = open(filename, 'r', encoding= 'utf-8')
        return f.readlines()
    except IOError:
        print("Error opening or reading input file: ", filename)
        sys.exit()

def get_words_from_line_list(L):
    """
    Parse the given list L of text lines into words.
    Return list of all words found.
    """
    word_list = []
    for line in L:
        words_in_line = get_words_from_string(line)
        word_list.extend(words_in_line)
    return word_list

def get_words_from_string(line):
    """
    Return a list of the words in the given input string,
    converting each word to lower-case.

    Input:  line (a string)
    Output: a list of strings 
              (each string is a sequence of alphanumeric characters)
    """
    line = line.translate(translation_table)
    word_list = line.split()
    return word_list

def count_frequency(word_list):
    """
    Input a list of words
    Return a DICTIONARY of (word, frequency) pairs
    """
    #######################
    #                     #
    #     TO IMPLEMENT    #
    #                     #
    #######################
    freq = {}
    for word in word_list:
        freq[word] = freq.get(word, 0) + 1
    return freq

def word_frequencies_for_file(filename):
    """
    Return dictionary of (word,frequency) pairs for the given file.
    """

    line_list = read_file(filename)
    word_list = get_words_from_line_list(line_list)
    freq_mapping = count_frequency(word_list)
    return freq_mapping

def inner_product(D1, D2):
    """
    Inner product between two vectors, where vectors
    are represented as dictionaries of (word,freq) pairs.

    Example: inner_product({"and":3,"of":2,"the":5},
                           {"and":4,"in":1,"of":1,"this":2}) = 14.0 
    """
    #######################
    #                     #
    #     TO IMPLEMENT    #
    #                     #
    #######################
    sum = 0.0
    if len(D1) < len(D2):
        for key in D1:
            if key in D2:
                sum += D1[key] * D2[key]
    else:
        for key in D2:
            if key in D1:
                sum += D1[key] * D2[key]
    return sum

def vector_angle(D1, D2):
    """
    The input are two vectors represented as dictionary of (word,freq) pairs.
    Return the angle between these two vectors.
    """
    numerator = inner_product(D1, D2)
    denominator = math.sqrt(inner_product(D1, D1) * inner_product(D2, D2))
    return math.acos(numerator / denominator)

def docdist5():
    word_dict_1 = word_frequencies_for_file(filename_1)
    word_dict_2 = word_frequencies_for_file(filename_2)
    distance = vector_angle(word_dict_1, word_dict_2)
    print("The distance between the documents is: %0.6f (radians)"%distance)

cProfile.run("docdist5()")

The distance between the documents is: 0.619328 (radians)
         216306 function calls in 0.040 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    0.006    0.003 4269496019.py:1(read_file)
        2    0.003    0.002    0.021    0.010 4269496019.py:13(get_words_from_line_list)
    23110    0.003    0.000    0.016    0.000 4269496019.py:24(get_words_from_string)
        2    0.008    0.004    0.013    0.006 4269496019.py:37(count_frequency)
        2    0.000    0.000    0.039    0.019 4269496019.py:52(word_frequencies_for_file)
        3    0.000    0.000    0.000    0.000 4269496019.py:62(inner_product)
        1    0.000    0.000    0.000    0.000 4269496019.py:86(vector_angle)
        1    0.001    0.001    0.040    0.040 4269496019.py:95(docdist5)
        1    0.000    0.000    0.040    0.040 <string>:1(<module>)
        2    0.000    0.000    0.000    0.000 codecs.py:260(__init__)
        